opentracing-spring-cloud/src/test/java/io/opentracing/contrib/spring/cloud/log/LoggingAutoConfigurationTest.java

```diff
@@ -0,0 +1,231 @@
+    }
+
+    @RequestMapping("/log-error")
+    public ContextData logError() {
+      log4j.error(LOG);
+      return contextData(null, true, Level.ERROR.toString());
+    }
+  }
+
+  @Autowired
+  private TestRestTemplate restTemplate;
+  @Autowired
+  private MockTracer mockTracer;
+
+  @After
+  public void after() {
+    mockTracer.reset();
+  }
+
+  @Test
+  public void testJcl() {
+    ResponseEntity<ContextData> response = restTemplate.getForEntity("/jcl", ContextData.class);
+    assertLogging(response.getBody());
+  }
+
+  @Test
+  public void testSlf4j() {
+    ResponseEntity<ContextData> response = restTemplate.getForEntity("/slf4j", ContextData.class);
+    assertLogging(response.getBody());
+  }
+
+  @Test
+  public void testJul() {
+    ResponseEntity<ContextData> response = restTemplate.getForEntity("/jul", ContextData.class);
+    assertLogging(response.getBody());
+  }
+
+  @Test
+  public void testLog4j() {
+    ResponseEntity<ContextData> response = restTemplate.getForEntity("/log4j", ContextData.class);
+    assertLogging(response.getBody());
+  }
+
+  @Test
+  public void testNoException() {
+    ResponseEntity<ContextData> response = restTemplate.getForEntity("/no-exception", ContextData.class);
+    assertLogging(response.getBody());
+  }
+
+  @Test
+  public void testError() {
+    ResponseEntity<ContextData> response = restTemplate.getForEntity("/log-error", ContextData.class);
+    assertLogging(response.getBody());
+  }
+
+  private void assertLogging(ContextData contextData) {
+    List<MockSpan> mockSpans = mockTracer.finishedSpans();
+    assertEquals(1, mockSpans.size());
+    MockSpan mockSpan = mockSpans.get(0);
+    // preHandle log
+    // standard log
+    // afterCompletion log
+    assertEquals(3, mockSpan.logEntries().size());
+    LogEntry logEntry = mockSpan.logEntries().get(1);
+
+    if (contextData.getThrowable() == null && !contextData.isError()) {
+      assertEquals(4, logEntry.fields().size());
+    } else if (contextData.getThrowable() != null && contextData.isError()) {
+      assertEquals(6, logEntry.fields().size());
+    } else {
+      assertEquals(5, logEntry.fields().size());
+    }
+
+
+    assertEquals(Controller.class.getName(), logEntry.fields().get("logger"));
+    assertEquals(LOG, logEntry.fields().get("message"));
+    assertEquals(contextData.getThread(), logEntry.fields().get("thread"));
+    assertEquals(contextData.getLevel(), logEntry.fields().get("level"));
+    if (contextData.getThrowable() != null) {
+      assertEquals(contextData.getThrowable().getMessage(), ((Throwable)logEntry.fields().get("error.object")).getMessage());
+    }
+    if (contextData.isError()) {
+      assertEquals(Tags.ERROR, logEntry.fields().get("event"));
+    }
+    // now >= timestamp < now + 10ms
+    assertTrue(contextData.getTimestamp() >= logEntry.timestampMicros()  && logEntry.timestampMicros() < contextData.getTimestamp() + 10 * 1000);
```

Comment:
Shouldn't the second part be `logEntry.timestampMicros() + 10*1000 > contextData.getTimestamp()`? - to make sure the different is no more than 10ms.

Reply:
that is better, I will update it